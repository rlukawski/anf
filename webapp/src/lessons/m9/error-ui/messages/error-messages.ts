// słownik komunikatów błędów
// wszystkie komunikaty, wyświetlane użytkownikom, są spisane tutaj
export const errorMessagesMap = {
  UNEXPECTED_ERROR: "Unexpected error occurred. Please contact Support.",
  CONNECTION_FAILURE: "Internet connection error. Check your wires!",
  PAID_CONTENT: "This content is paid. You have to upgrade to access it.",
  FORBIDDEN: "Access to this resource is forbidden.",
  NOT_FOUND: "Broken link. The page you're trying to access doesn't exist.",
  TIMEOUT: "This resource is no longer available.",
  SYSTEM_FAILURE: "Something is wrong, but we're working on it!",
}

// KOMUNIKAT BŁĘDU jako taki - dostaje swój typ
// w razie potrzeby w przyszłości można go rozszerzać
export interface UserErrorMessage {
  type: keyof typeof errorMessagesMap
}

// alternatywnie, po taniości można by dać:
// export type UserErrorMessage = string

//////////////////////////////////////////

// jeśli nasze API, kiedy rzuca błędami typu 401, 403, 500 etc. dodaje payload
// to - siłą rzeczy - powinno to mieć swój typ
// nie wszystkie API dodają payload do błędu, ale niektóre tak ;)
export interface APIError {
  id: string
  code: string
  message: string
}

//////////////////////////////////////////

// poniższy kodzik mapuje statusy błędów HTTP na konkretne komunikaty
// ta funkcja jest wykorzystywana wewn. getUserErrorMessage
// i robi mniej więcej tyle, że jeśli leci błąd 403, to dla niego bierzemy komunikat "FORBIDDEN"
// a "FORBIDDEN" jest zdefiniowany w słowniku na górze

// znowu - w zależności od wymagań - niektóre systemy będą wyświetlały konkretne komunikaty błędów w zależności od statusu HTTP - a inne nie
// my tutaj to obsługujemy.
// w efekcie, jeśli gdziekolwiek w aplikacji poleci np. 403 i używamy `getUserErrorMessage`, to poleci ten sam komunikat.
// ale JEŚLI tego apka NIE POTRZEBUJE - to nie ma sensu robić NADMIAROWEGO kodu

export const getErrorMessage = (errMessage: UserErrorMessage) =>
  errorMessagesMap[errMessage.type]

export const HTTPStatusCodeToErrorMessageMap: {
  [code in number]?: keyof typeof errorMessagesMap;
} = {
  403: "FORBIDDEN",
  404: "NOT_FOUND",
  410: "TIMEOUT",
};

export const getErrorMessageByHTTPStatusCode = (statusCode: number) => {
  const result = HTTPStatusCodeToErrorMessageMap[statusCode];
  return result;
};
