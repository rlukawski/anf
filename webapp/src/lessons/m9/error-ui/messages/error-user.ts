import { AxiosError } from "axios";

import * as guards from "./error-guards";
import { UserErrorMessage, getErrorMessageByHTTPStatusCode } from "./error-messages";

const getErrorMessageIfAxiosError = (e: AxiosError) => {
  if (e.response?.status) {
    return getErrorMessageByHTTPStatusCode(e.response.status);
  }
};

// kaskada IFów. Tę funkcję możemy wpiąć w try..catche w aplikacji
export const getUserErrorMessage = (e: unknown): UserErrorMessage => {
  // najpierw sprawdzamy czy rzucony został nie-error
  // (nie spodziewamy się tego, ale w JSie niestety to możliwe)
  // e: unknown
  if (!(e instanceof Error)) {
    return { type: "UNEXPECTED_ERROR" };
  }

  // potem sprawdzamy type-errory
  // czyli potencjalnie "lipa w naszym kodzie"
  // e: TypeError
  if (e instanceof TypeError) {
    return { type: "UNEXPECTED_ERROR" };
  }

  // następnie sprawdzamy wszelkie inne errory, byle NIE axios (nie HTTP)
  // e: Error
  if (!guards.isAxiosError(e)) {
    return { type: "UNEXPECTED_ERROR" };
  }

  // 🔥 widzimy, że mamy 3 IFy, które mapują na ten sam komunikat
  // 🔥 czy lepiej mieć je osobno - dla czytelności - czy połączyć - KWESTIA PREFERENCJI
  // 🔥 w naszym repo są wydzielone, aby łatwiej było śledzić, co się dzieje

  // a teraz - brak internetu
  // Axios Network Error (see https://github.com/axios/axios/issues/383)
  // thrown when Axios is unable to receive any response from the server, as a consequence both status and response are undefined
  // Od wersji 0.27.0 błąd braku sieci ma kod 'ERR_NETWORK' (https://github.com/axios/axios/pull/3645)
  if (!e.response || e.code === "ERR_NETWORK") {
    return { type: "CONNECTION_FAILURE" };
  }

  // a od tego miejsca, skoro obsłużyliśmy wszelkie "egzotyczne" błędy
  // to przechodzimy do błędów HTTP. Mamy `response`, więc MUSIELIŚMY DOSTAĆ ODPOWIEDŹ
  // no i teraz - w taki albo śmaki sposób - mapujemy statusy HTTP na konkretne komunikaty
  // poniższy kod to tylko przykład, można go zamodelować na wiele sposobów

  // AxiosError, use HTTP Status Code map
  const messageType = getErrorMessageIfAxiosError(e);
  // const messageType = getErrorMessageByHTTPStatusCode(parseInt(e.code))
  if (messageType) {
    return { type: messageType };
  }

  // Axios Error 5xx
  if (guards.isServerError(e)) {
    return { type: "SYSTEM_FAILURE" };
  }

  return { type: "UNEXPECTED_ERROR" };
}
