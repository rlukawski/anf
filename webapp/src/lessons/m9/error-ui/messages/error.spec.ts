import axios from "axios";
import { rest } from "msw";
import { setupServer } from "msw/node";

import { getUserErrorMessage } from "./error-user";
import { HTTPStatusCodeToErrorMessageMap } from "./error-messages";

const handlers = [
  rest.get("/mocks/errors/network", async (req, res, ctx) => {
    return res.networkError("Failed to connect");
  }),
  rest.get(`/mocks/errors/403`, async (req, res, ctx) => {
    return res(ctx.status(403, `Backend error message 403`));
  }),
  rest.get(`/mocks/errors/404`, async (req, res, ctx) => {
    return res(ctx.status(404, `Backend error message 404`));
  }),
  rest.get(`/mocks/errors/410`, async (req, res, ctx) => {
    return res(ctx.status(410, `Backend error message 410`));
  }),
];
const server = setupServer(...handlers);

describe("getUserErrorMessage", () => {
  beforeAll(() =>
    server.listen({
      onUnhandledRequest: "warn",
    })
  );
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  it("should produce UNEXPECTED_ERROR if a non-error was thrown", () => {
    expect(getUserErrorMessage(1)).toMatchInlineSnapshot(`
      Object {
        "type": "UNEXPECTED_ERROR",
      }
    `);
  });

  it("should produce UNEXPECTED_ERROR if a TypeError was thrown", () => {
    const error = new TypeError("undefined is not a function");
    expect(getUserErrorMessage(error)).toMatchInlineSnapshot(`
      Object {
        "type": "UNEXPECTED_ERROR",
      }
    `);
  });

  it("should produce CONNECTION_FAILURE if a HTTP NetworkError was thrown", async () => {
    try {
      await axios.get("/mocks/errors/network");
      fail(
        "error should be thrown, if it's not then probably HTTP is misconfigured"
      );
    } catch (error: unknown) {
      expect(getUserErrorMessage(error)).toMatchInlineSnapshot(`
        Object {
          "type": "CONNECTION_FAILURE",
        }
      `);
    }
  });

  const testcases = Object.entries(HTTPStatusCodeToErrorMessageMap)
    .map(([code, status]) => [status, parseInt(code)] as const)

  it.each(testcases)(`should produce %s if a HTTP %s was thrown`, async (errorMessage, statusCode) => {
    try {
      await axios.get(`/mocks/errors/${statusCode}`);
      fail("error should be thrown, if it's not then probably HTTP is misconfigured");
    } catch (error: unknown) {
      // Jest: Multiple inline snapshots for the same call are not supported. :(
      expect(getUserErrorMessage(error)).toMatchSnapshot();
    }
  })
});
