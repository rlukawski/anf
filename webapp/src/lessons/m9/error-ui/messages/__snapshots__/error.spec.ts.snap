// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserErrorMessage should produce FORBIDDEN if a HTTP 403 was thrown 1`] = `
Object {
  "type": "FORBIDDEN",
}
`;

exports[`getUserErrorMessage should produce NOT_FOUND if a HTTP 404 was thrown 1`] = `
Object {
  "type": "NOT_FOUND",
}
`;

exports[`getUserErrorMessage should produce TIMEOUT if a HTTP 410 was thrown 1`] = `
Object {
  "type": "TIMEOUT",
}
`;
